src/plugins/tscn/tscnplugin.cpp

```diff
@@ -0,0 +1,666 @@
+/*
+ * Godot 4 Scene Tiled Plugin
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "tscnplugin.h"
+
+#include "grouplayer.h"
+#include "map.h"
+#include "mapobject.h"
+#include "objectgroup.h"
+#include "savefile.h"
+#include "tile.h"
+#include "tilelayer.h"
+
+#include <QCoreApplication>
+#include <QDir>
+#include <QDirIterator>
+#include <QFileInfo>
+#include <QMap>
+#include <QRegularExpression>
+
+#include <iostream>
+#include <stdexcept>
+
+using namespace Tiled;
+using namespace Tscn;
+
+TscnPlugin::TscnPlugin()
+{
+}
+
+// Convenience functions for throwing translated errors
+static std::invalid_argument tscnError(const char* error)
+{
+    return std::invalid_argument(QCoreApplication::translate("TscnPlugin", error).toStdString());
+}
+
+template <typename... Types>
+static std::invalid_argument tscnError(const char* error, Types... args)
+{
+    return std::invalid_argument(
+        QCoreApplication::translate("TscnPlugin", error).arg(args...).toStdString()
+    );
+}
+
+/**
+ * Determines the res:// root for the Godot project. This is done by searching
+ * for a .godot file in the parent paths of the file.
+ * If there's no .godot file found in a shared root, it returns a blank string.
+ */
+static QString determineResRoot(const QString &filePath)
+{
+    constexpr unsigned int searchDepth = 3;
+    QFileInfo fi(filePath);
+    QDir dir(fi.path());
+    dir.setNameFilters(QStringList("*.godot"));
+
+    for (unsigned int i = 0; i < searchDepth; ++i) {
+        if (i > 0 && !dir.cdUp())
+            break;
+
+        const QString godotFile = QDirIterator(dir).next();
+        if (!godotFile.isEmpty()) {
+            return dir.path();
+        }
+    }
+
+    throw tscnError("Could not find .godot project in file path for file %1", filePath);
+}
+
+// Converts a tileset's image source to a res:// path
+// If resRoot is null, this will find the nearest .godot project and set that to the resRoot,
+// otherwise, it will ensure that the image source is a child of resRoot.
+static QString imageSourceToRes(const Tileset* tileset, std::unique_ptr<QString>& resRoot)
```

Comment:
Rather than using a `std::unique_ptr`, could `resRoot` being an empty string be enough to indicate its unset state? If not, since we're using C++17 now, I'd suggest using an `std::optional` instead.

Reply:
Good call; I can't think of a reason why an empty resRoot would ever be a valid value. Also nice to know you're on C++17.